File: Exclusive_Time.py
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None


// Your code here along with comments explaining your approach:
In this problem we use a stack and if the function is start then we make previous pointer to curr pointer and then push it to the stack and if the function is end then we change prev to curr+1 and then pop it from the stack.


# Time complexity --> o(n)
# space complexity --> o(n)
class Solution(object):
    def exclusiveTime(self, n, logs):
        """
        :type n: int
        :type logs: List[str]
        :rtype: List[int]
        """
        prev=0
        curr=0
        stack=[]
        res=[0 for i in range(n)]
        for i in range(len(logs)):
            log=logs[i]
            val=log.split(':')
            functionid=int(val[0])
            curr=int(val[2])
            #when the function is start then we push it to the stack and we manage curr and prev.
            if val[1]=='start':
                if len(stack)>0:
                    res[stack[-1]]=res[stack[-1]]+curr-prev
                    prev=curr
                stack.append(functionid)
            #If the fucntion is end then we manage curr and prev and pop the element from the stack.
            else:
                res[functionid]=res[functionid]+curr-prev+1
                prev=curr+1
                stack.pop()
        return res
            

File: exclusiveTime.js
/**
 * @param {number} n
 * @param {string[]} logs
 * @return {number[]}
 */
var exclusiveTime = function(n, logs) {
    //base
    if(logs == null || logs.length == 0) return [];
    let prev = 0;
    let result = new Array(n).fill(0);
    let st = [];
    for(let i = 0; i< logs.length; i++){
        let splitStr = logs[i].split(':');
        let curr = parseInt(splitStr[2]);
        
        //start
        if(splitStr[1] == 'start'){
            //calculate the time lapsed so far for the current task(splitStr[0]), if in stack
            if(st.length != 0){
              let j = st[st.length -1];
               result[j] +=  curr - prev;
            }
            //push the current task into stack
            st.push(splitStr[0]);
            prev = curr; (prev = 2)
            
        } else {
            let j = st.pop();
            result[j] +=  curr - prev + 1;
            prev = curr + 1; (prev = 5 - 2 +1 )
        }
    }
    
    return result;
};
//time coplexity : O(n)